### PR TITLE
Fix hashing when passing an object with circular dependencies to an ApiCall as a parameter

### PR DESCRIPTION
**Pull request by bengry**
Currently Paris uses `JSON.stringify` as an object hash when making usage of `ApiCall`.
This doesn't allow abstraction of parameter extraction from an entity, for example:
```typescript
@Entity({
  ...
})
export class Todo extends EntityModelBase<string> {
  ...
}

@ApiCall({
  ...
  endpoint: (config, query) => `todos/${todoId}/comment`
  parseQuery: (todo: Todo) => ({ todoId: todo.id })
})
export class TodoCommentApiCall extends ApiCallModel<..., Todo> {}
```

since `Todo` is an `EntityModelBase` it has a `$parent` property which may be a circular reference.

In this case we can change the parameters of `TodoCommentApiCall` to be `{ todoId: Todo['id'] }`, but if you have more than a few things you need from the entity, this is very cumbersome, and is prone to code duplication in the application (each usage needs to extract the parameters from the entity before handing off to Paris).

The solution is to use a more safe approach to hashing, namely [`json-stringify-safe`](https://www.npmjs.com/package/json-stringify-safe), to create the same mechanism for creating the keys, but stop at circular references (which `JSON.strinfigy` doesn't do by default).

No breaking changes expected, apart from adding another `peerDependency` (this can be mitigated with adding `object-hash` as a `bundleDependencies`, but I'd like to avoid it if possible, if only to reduce the chance of duplicated libraries in different versions at runtime).
Note that this is an `optionalDependencies` since if you don't use caching or `ApiCall`s at all - this is unneeded.

---
An earlier attempt was made at this at #7, which used [`object-hash`](https://www.npmjs.com/package/object-hash), but was dropped [due to](https://github.com/Microsoft/paris/pull/7#issuecomment-414119439):
> I found an issue where object-hash will evaluate getter functions, which should not be part of the hash really, since they're essentially a computed value on top of the raw data.
I'll change the implementation to a regular JSON.stringify (which does not do the above) and pass a replacer to deal with cycllic dependencies and send another PR for that, since it changes most of the code in this one.

### CODE REVIEW

File: lib/repository/readonly-repository.ts [hunk 1]
```diff
@@ -452,7 +452,8 @@ export class ReadonlyRepository<TEntity extends ModelBase, TRawData = any> imple
 			model$ = of(model);
 		}
 
-		return entity.readonly ? model$.pipe(map(model => Object.freeze(model))) : model$;
+		// typecasting is used here until bug in TS is fixed (https://github.com/Microsoft/TypeScript/issues/26495)
+		return (entity.readonly ? model$.pipe(map(model => Object.freeze(model))) : model$) as TEntity extends TConcreteEntity ? Observable<TEntity> : Observable<TConcreteEntity>;
```
Comment (yoxigen) [inline lib/repository/readonly-repository.ts]: Seems like two separate issues. Shouldn't they be two separate PRs?

File: lib/services/paris.ts [hunk 2]
```diff
@@ -143,7 +144,7 @@ export class Paris{
 	 * @returns {Observable<TResult>} An Observable of the api call's result data type
 	 */
 	apiCall<TResult = any, TInput = any>(apiCallType:ApiCallType<TResult, TInput>, input?:TInput, dataOptions:DataOptions = defaultDataOptions):Observable<TResult>{
-		const cacheKey:string = JSON.stringify(input) || "{}";
+		const cacheKey:string = hashObject(input || {});
```
Comment (yoxigen) [inline lib/services/paris.ts]: A better solution might be to allow to pass a serialization function for the cache key, so it's not tied to Paris itself. That way you fix this problem, while also improving functionality in that the user can decide to cache together different calls (if there's a param that isn't relevant to caching, for example).

Something like `getCacheKey` on the ApiCall configuration, maybe?

And of course, this avoids adding another peer dependency to Paris, which adds complexity for its users.

File: lib/repository/readonly-repository.ts (hunk 1 above)
Comment (bengry) [reply lib/repository/readonly-repository.ts]: They are. See #6. It'll be completed before this one.

File: lib/services/paris.ts (hunk 2 above)
Comment (bengry) [reply lib/services/paris.ts]: I think getting this as a parameter is somewhat of an overkill, for a few reasons.
1. The way serialization works is 100% internal to Paris in this case, it's not like you (as a Paris user) ever get that cache back. You to create the cache key and never see it again. If we would to supply a whole caching layer from the outside instead, might've been more useful than just supplying the key.
  1.1. If we do pass it from the outside, it should probably be as part of the `ParisConfig`, since you'll pretty likely to pass the same implementation every time. There's no real reason to use different caching mechanisms for different entities. Moreover, it's dangerous - since different functions can return the same values for different inputs, causing a collision internally.
2. It's an `optionalDependency` - if you don't ever need caching - you don't need this.
  2.1. We can put it as a [`bundledDependencies`](https://npm.github.io/using-pkgs-docs/package-json/types/bundleddependencies.html) instead, this way it doesn't need to be part of the app's direct dependencies, but still there.

Also, I think the code above is outdated, I was originally using [`object-hash`](https://www.npmjs.com/package/object-hash) to do that, but it evaluates getters, which is an unwanted side-effect in this case. Changed to use [`json-stringify-safe`](https://www.npmjs.com/package/json-stringify-safe) instead, which is just `JSON.stringify` wrapped in a safeguard for circular dependencies. I'll check why the code above shows up and update this comment.

File: lib/repository/readonly-repository.ts (hunk 1 above)
Comment (bengry) [reply lib/repository/readonly-repository.ts]: #6 completed, this change is no longer part of this PR (it was added just so I can build this and test it out).
